File: day02/part2.lua
local function is_monotonic(lst)
    local ascending = true
    local descending = true
    
    for i = 2, #lst do
        if lst[i] < lst[i-1] then
            ascending = false
        end
        if lst[i] > lst[i-1] then
            descending = false
        end
    end
    
    return ascending or descending
end

local function valid_differences(lst)
    for i = 2, #lst do
        local diff = math.abs(lst[i] - lst[i-1])
        if diff < 1 or diff > 3 then
            return false
        end
    end
    return true
end

local function safe_report(report)
    return is_monotonic(report) and valid_differences(report)
end

local function parse_reports(input_string)
    local reports = {}
    for line in input_string:gmatch("[^\r\n]+") do
        local report = {}
        for num in line:gmatch("%d+") do
            table.insert(report, tonumber(num))
        end
        table.insert(reports, report)
    end
    return reports
end

local function safe_report_with_dampener(report)
    -- Check if the report is safe without removing any level
    if safe_report(report) then
        return true
    end
    
    -- Try removing each level and check if the resulting list is safe
    for i = 1, #report do
        local modified_report = {}
        for j = 1, #report do
            if j ~= i then
                table.insert(modified_report, report[j])
            end
        end
        
        if safe_report(modified_report) then
            return true
        end
    end
    
    return false
end

local function count_safe_reports_with_dampener(reports)
    local count = 0
    for _, report in ipairs(reports) do
        if safe_report_with_dampener(report) then
            count = count + 1
        end
    end
    return count
end

local function solve(filename)
    local file = io.open(filename, "r")
    if not file then return 0 end
    
    local input_string = file:read("*all")
    file:close()
    
    local reports = parse_reports(input_string)
    return count_safe_reports_with_dampener(reports)
end

print(solve("input.txt"))

File: day02/part2.exs
defmodule Part2 do
  defp monotonic?(lst) do
    Enum.sort(lst) == lst || Enum.sort(lst, :desc) == lst
  end

  defp valid_differences?(lst) do
    lst
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.all?(fn [a, b] ->
      diff = abs(b - a)
      diff >= 1 && diff <= 3
    end)
  end

  defp safe_report?(report) do
    monotonic?(report) && valid_differences?(report)
  end

  defp safe_report_with_dampener?(report) do
    # Check if the report is safe without removing any level
    if safe_report?(report) do
      true
    else
      # Try removing each level and check if the resulting list is safe
      Enum.with_index(report)
      |> Enum.any?(fn {_, index} ->
        modified_report = List.delete_at(report, index)
        monotonic?(modified_report) && valid_differences?(modified_report)
      end)
    end
  end

  defp parse_reports(input_string) do
    input_string
    |> String.split("\n", trim: true)
    |> Enum.map(fn report_str ->
      report_str
      |> String.split()
      |> Enum.map(&String.to_integer/1)
    end)
  end

  defp count_safe_reports_with_dampener(reports) do
    reports
    |> Enum.filter(&safe_report_with_dampener?/1)
    |> Enum.count()
  end

  def solve(filename) do
    filename
    |> File.read!()
    |> parse_reports()
    |> count_safe_reports_with_dampener()
  end
end

"input.txt"
|> Part2.solve()
|> IO.puts()
